File: cpp/ex2_3.cpp
#include <functional>

using namespace std;

class Person {};
class Student : public Person {};

void startMentoring(function<Person *(Student *)> select) {
    Student *st = new Student();
    Person *mentor = select(st);
}

Student *selectStudentMentor(Student *st) { return new Student(); }

void sendEmails(function<bool(Student *)> needEmail) {
    Student *st = new Student();
    if (needEmail(st)) {}
}

bool isHacked(Person *pr) { return true; }

int main() {
    startMentoring(selectStudentMentor);
    sendEmails(isHacked);
}

File: typescript/ex2_3.ts
class Person { email: string = ""; }
class Student { email: string = ""; grade: number = 0; }

function startMentoring(select: (s: Student) => Person): void {
    let st: Student = new Student();
    let mentor: Person = select(st);
}

function selectStudentMentor(st: Student): Student { return new Student(); }
startMentoring(selectStudentMentor);

function sendEmails(needEmail: (s: Student) => Boolean): void {
    let st: Student = new Student();
    if (needEmail(st)) {}
}

function isHacked(pr: Person): Boolean { return true; }
sendEmails(isHacked);
